File: packages/oxlint-plugin-react-doctor/src/plugin/rules/security-scan/firebase-query-filter-as-auth.ts
import { defineRule } from "../../utils/define-rule.js";
import { isClientSourcePath } from "./utils/is-client-source-path.js";
import { scanByPattern } from "./utils/scan-by-pattern.js";

export const firebaseQueryFilterAsAuth = defineRule({
  id: "firebase-query-filter-as-auth",
  title: "Firestore query filter used as authorization",
  severity: "warn",
  impact: "security",
  confidence: "heuristic",
  fix: "local",
  recommendation:
    "Make sure Firestore rules compare the requested document against `request.auth.uid` and trusted membership data.",
  scan: scanByPattern({
    shouldScan: (file) => isClientSourcePath(file.relativePath),
    pattern:
      /\.where\s*\(\s*["'](?:uid|userId|userID|ownerId|ownerID|orgId|orgID|tenantId|tenantID|role)["']\s*,\s*["']==["']/i,
    message:
      "Firestore query code filters by an auth-shaped field; filtering is not authorization unless rules enforce the same boundary.",
  }),
});

File: native/oxlint/scans/firebase-query-filter-as-auth.rs
use lazy_regex::{Lazy, Regex, lazy_regex};

use super::{
    ScanFinding, first_pattern_finding::first_pattern_finding, scan_content::ScanContent,
};

const MESSAGE: &str = "Firestore query code filters by an auth-shaped field; filtering is not authorization unless rules enforce the same boundary.";

static QUERY_FILTER_PATTERN: Lazy<Regex> = lazy_regex!(
    r#"(?i)\.where\s*\(\s*["'](?:uid|userId|userID|ownerId|ownerID|orgId|orgID|tenantId|tenantID|role)["']\s*,\s*["']==["']"#
);

pub fn scan(relative_path: &str, source: &ScanContent<'_>) -> Vec<ScanFinding> {
    if !super::is_client_source_path::is_client_source_path(relative_path) {
        return Vec::new();
    }
    let content = source.normalized_scannable(false);
    first_pattern_finding(source, &content, &[&QUERY_FILTER_PATTERN], MESSAGE)
}
